[DOCU-3062] Composite runtime groups (#5762)
Kong/docs.konghq.com: app/konnect/runtime-manager/composite-runtime-groups/migrate.md

---
title: Migrate configuration into composite runtime groups
content_type: how-to
badge: enterprise
---

There are a few situations in which you might want to migrate existing configuration into a composite runtime group.

> * Migrate standard runtime group configuration into a composite group. In this scenario, you are migrating config from Konnect to a different group in Konnect.

Reviewer: ```suggestion
* Migrate standard runtime group configuration into a composite group. In this scenario, you are migrating config from {{site.konnect_short_name}} to a different group in {{site.konnect_short_name}}.
```

> * Migrate self-managed Kong Gateway workspace configuration into a composite group in Konnect. In this scenario, you are migrating from self-managed to cloud managed.

Reviewer: ```suggestion
* Migrate self-managed {{site.base_gateway}} workspace configuration into a composite group in {{site.konnect_short_name}}. In this scenario, you are migrating from self-managed to cloud managed.
```

        -H "Authorization: Bearer <your_KPAT>" \
        --data name=CRG \
        --data cluster_type=CLUSTER_TYPE_COMPOSITE
    ```

1. Create a new standard runtime group:

    ```sh
    curl -i -X POST https://<region>.api.konghq.com/v2/runtime-groups \
        -H "Authorization: Bearer <your_KPAT>" \
        --data name=SRG1 \
        --data cluster_type=CLUSTER_TYPE_HYBRID
    ```

1. Add the new group to the composite runtime group as a member:

    ```sh
    curl -i -X POST https://<region>.api.konghq.com/v2/runtime-groups/<composite-group-ID>/composite-memberships/add \
        -H "Authorization: Bearer <your_KPAT>" \
        --json '{"members": [{"id": "062e2f2c-0f42-4938-91b4-f73f399260f5"}]}'
    ```

    {:.note}
    > **Note**: When adding a standard group to a composite, make sure it has no connected 
    runtime instances.

## Scenario: Migrate runtime group configuration into a composite group

If you want to migrate a standard runtime group into a composite, Kong recommends the following path:

1. Create a composite runtime group.
1. Create a new standard runtime group and add as a member to the composite group.

> 1. Dump configuration from old group.

Reviewer 2: I interpreted this as "dump config from old composite RG instead of standard RG". Can we use some example names to help users understand the migration better?

> 1. Dump configuration from old group.

Reviewer 3: ```suggestion
1. Dump the configuration from the old standard runtime group.
```

> 1. Apply configuration to new member group.

Reviewer 3: ```suggestion
1. Apply configuration to new standard runtime group that is a member of the composite runtime group.
```

---
title: Migrate configuration into composite runtime groups
content_type: how-to
badge: enterprise
---

There are a few situations in which you might want to migrate existing configuration into a composite runtime group.

* Migrate standard runtime group configuration into a composite group. In this scenario, you are migrating config from Konnect to a different group in Konnect.
* Migrate self-managed Kong Gateway workspace configuration into a composite group in Konnect. In this scenario, you are migrating from self-managed to cloud managed.

> A composite runtime group **can't be configured directly**. It compiles configuration from its child runtime groups.

Reviewer 4: question: Are we using parent/child terminology? If not, I suggest:

```suggestion
A composite runtime group **can't be configured directly**. It compiles configuration from its member runtime groups.
```

## Prepare runtime groups for migration

1. Create a composite runtime group:

    ```sh
    curl -i -X POST https://<region>.api.konghq.com/v2/runtime-groups \
        -H "Authorization: Bearer <your_KPAT>" \
        --data name=CRG \
        --data cluster_type=CLUSTER_TYPE_COMPOSITE
    ```

1. Create a new standard runtime group:

    ```sh
    curl -i -X POST https://<region>.api.konghq.com/v2/runtime-groups \
        -H "Authorization: Bearer <your_KPAT>" \
        --data name=SRG1 \
        --data cluster_type=CLUSTER_TYPE_HYBRID
    ```

1. Add the new group to the composite runtime group as a member:

    ```sh
    curl -i -X POST https://<region>.api.konghq.com/v2/runtime-groups/<composite-group-ID>/composite-memberships/add \
        -H "Authorization: Bearer <your_KPAT>" \
        --json '{"members": [{"id": "062e2f2c-0f42-4938-91b4-f73f399260f5"}]}'
    ```

    When adding a standard group to a composite, make sure it has no connected runtime instances.

    A composite runtime group can contain up to 256 runtime groups. 
    You can add or remove up to 50 child runtime groups at a time.

>     Each standard group can have up to 5 parent composite runtime groups.

Reviewer 4: nitpick: This is duplicated documentation that also exists on other pages. Do we want to repeat information? Also, use of "child/parent"?

Author: Will remove this one, since it's not really relevant when you're adding one runtime group.

## Scenario: Migrate runtime group configuration into a composite group

If you want to migrate a standard runtime group into a composite, Kong recommends the following path:

1. Create a composite runtime group.
1. Create a new standard runtime group and add as a member to the composite group.
1. Dump configuration from old group.
1. Apply configuration to new member group.
1. Start new runtime instances in composite runtime group and test routes.
1. Decommission the old standard runtime group.

We recommend this workflow because it gives you the opportunity to review shared entities and avoid conflicts. 
It also decouples any team, dev portal, or identity permissions from the group. This way, you won't accidentally 
grant access to the wrong resources to a new group of users.

> Assuming you already have a composite runtime group and a child standard runtime group, you can export configuration from the old group and apply it to the new one.

Reviewer 4: question: Are we using parent/child terminology? If not, I suggest:

```suggestion
Assuming you already have a composite runtime group and a member runtime group, you can export configuration from the old group and apply it to the new one.
```

1. Export the configuration of the old runtime group via `deck dump`:

    ```sh
    deck dump \
        -o old-group.yaml \
        --konnect-token <your_KPAT> \
        --konnect-runtime-group-name old-group
    ```

1. Sync the configuration to the new group:

    ```sh
    deck sync \
        -s old-group.yaml \
        --konnect-token <your_KPAT> \
        --konnect-runtime-group-name SRG1
    ```

>     Note that you can't sync the configuration to the composite group `CRG`. Composite runtime groups don't have their own configuration. They use combined configuration from all of their child groups.

Reviewer 4: question: Are we using parent/child terminology? If not, I suggest:
```suggestion
    Note that you can't sync the configuration to the composite group `CRG`. Composite runtime groups don't have their own configuration. They use combined configuration from all of their member runtime groups.
```
